### PR TITLE
heal items

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
also: improve logging for archive check

https://paste.thinkmo.de/MMXNxFUh#borg-healing.txt

### CODE REVIEW

File: borg/archive.py [hunk 1]
```diff
@@ -912,31 +912,55 @@ def add_reference(id_, size, csize, cdata=None):
...
+                        logger.info('{}: Healed previously missing file chunk! (Byte {}-{}).'.format(
+                            item[b'path'].decode('utf-8', 'surrogateescape'), offset, offset + size))
+                        add_reference(chunk_id, size, csize)
                 chunk_list.append((chunk_id, size, csize))
                 offset += size
-            if chunks_replaced and b'chunks_healthy' not in item:
+            if chunks_replaced and not has_chunks_healthy:
                 # if this is first repair, remember the correct chunk IDs, so we can maybe heal the file later
                 item[b'chunks_healthy'] = item[b'chunks']
+            if has_chunks_healthy and chunk_list == chunks_healthy:
```
Comment (ThomasWaldmann) [inline borg/archive.py]: somehow condition does not trigger.

**Comment by enkore** [comment]
Hm, I'd really like a test for this logic. Could dispose of chunks by just deleting them in the test. Check that the file get's the zero chunk, add chunk back to repository in the tests, check that it's repaired.

And another test showing that successive check--repair's with more and more missing chunks doesn't throw the original chunk ID list away?

**Comment by ThomasWaldmann** [comment]
Yeah, I'll try to add reasonable test and docs.

**Comment by ThomasWaldmann** [comment]
merge?

**Pull request merged**